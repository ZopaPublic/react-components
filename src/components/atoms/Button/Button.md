### Summary

Use `<Button />` whenever you need to render a call to action within a **Zopa interface**.

It comes in a lot of variations to suit different needs.

### Examples

- Primary ( _used for the main call to action, should only appear once per screen._ )

```tsx
import { Button } from '@zopauk/react-components';

<Button>Primary Button</Button>;
```

- Secondary ( _standard button for most actions_ )

```tsx
import { Button } from '@zopauk/react-components';

<Button styling="secondary">Secondary Button</Button>;
```

- Disabled state ( _for use when actions are disabled, both primary and secondary_ )

```tsx
import { Button } from '@zopauk/react-components';

<Button disabled={true}>Disabled Button</Button>;
```

- Text flavour ( _used to navigate to other pages, always in current window unless external icon is present_ )

```tsx
import { Button } from '@zopauk/react-components';

<Button styling="link">Text Button</Button>;
```

- As wide as its container

```tsx
import { Button, FlexRow, FlexCol, FlexContainer } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={8}>
      <Button fullWidth>Wide Button</Button>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Loading ( _spinner won't show for `styling="link"`, loading buttons are always `disabled`_ )

```tsx
import { Button, FlexRow, FlexCol, FlexContainer } from '@zopauk/react-components';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={6}>
      <Button loading>Loading</Button>
    </FlexCol>
    <FlexCol xs={6}>
      <Button styling="secondary" loading>
        Waiting
      </Button>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- With icons

```tsx
import { Button, FlexRow, FlexCol, FlexContainer } from '@zopauk/react-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowCircleRight, faArrowCircleLeft, faCoffee } from '@fortawesome/free-solid-svg-icons';

<FlexContainer>
  <FlexRow>
    <FlexCol xs={4}>
      <Button>
        <FontAwesomeIcon icon={faArrowCircleLeft} />
        {'\u00A0'} Previous
      </Button>
    </FlexCol>
    <FlexCol xs={4}>
      <Button>
        Have some {'\u00A0'}
        <FontAwesomeIcon icon={faCoffee} />
      </Button>
    </FlexCol>
    <FlexCol xs={4}>
      <Button>
        Next {'\u00A0'}
        <FontAwesomeIcon icon={faArrowCircleRight} />
      </Button>
    </FlexCol>
  </FlexRow>
</FlexContainer>;
```

- Button as a custom component

```tsx
import React from 'react';
import styled from 'styled-components';
import { buttonStyle } from '@zopauk/react-components';

// could be a gatsby or react-router-dom <Link />
const Link = ({ href, ...rest }) => <a href={href} {...rest} />;

const ButtonLink = styled(Link)`
  ${buttonStyle}
`;

<ButtonLink href="https://www.zopa.com" styling="secondary">
  Custom Button
</ButtonLink>;
```

- Disabled anchor element with button style

```tsx
import { buttonStyle, ButtonProps } from '@zopauk/react-components';
import { MouseEvent } from 'react';
import styled from 'styled-components';

const StyledA = styled.a`
  ${buttonStyle}
`;

const ButtonLink = ({ href, styling, fullWidth, children, disabled }) => (
  <StyledA
    href={href}
    styling={styling}
    fullWidth={fullWidth}
    disabled={disabled}
    onClick={(e) => disabled && e.preventDefault()}
  >
    {children}
  </StyledA>
);

<ButtonLink href="https://www.zopa.com" disabled>
  Button Link
</ButtonLink>;
```
