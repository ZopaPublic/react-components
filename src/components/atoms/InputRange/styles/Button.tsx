import React from 'react';
import styled from 'styled-components';
import grid from '../../../../constants/grid';
import { IButtonProps, buttonStyle } from '../../Button/Button';

const ButtonWrapper = styled.div`
  padding: 12px;
  @media (min-width: ${grid.breakpoints.m}px) {
    padding: 16px;
  }
`;

const SButton = styled.button`
  ${buttonStyle};
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  width: 32px;
  height: 32px;
  padding: 0;

  @media (min-width: ${grid.breakpoints.m}px) {
    width: 50px;
    height: 50px;
  }
`;

export const Button = (props: IButtonProps) => {
  return (
    <ButtonWrapper>
      <SButton {...props} />
    </ButtonWrapper>
  );
};
