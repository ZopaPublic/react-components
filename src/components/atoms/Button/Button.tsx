import Spinner from '../Spinner/Spinner';
import styled, { css } from 'styled-components';
import React, { ButtonHTMLAttributes } from 'react';
import { AppTheme, useThemeContext } from '../../styles/Theme';
import { colors, spacing, typography } from '../../../constants';

export type Styling = 'primary' | 'secondary' | 'link';

type BaseButtonProps = {
  styling?: Styling;
  loading?: boolean;
  disabled?: boolean;
  fullWidth?: boolean;
};

export type ButtonProps<T = ButtonHTMLAttributes<HTMLButtonElement>> = BaseButtonProps & T;

const colorMap = {
  primary: {
    text: colors.white,
    bg: colors.action,
    hover: `linear-gradient(90deg, #3B46C4 0%, #2732B0 100%)`,
  },
  secondary: {
    text: colors.actionDark,
    bg: colors.actionLight,
    hover: '#EEEFFB',
  },
  link: {
    text: colors.actionDark,
    bg: 'transparent',
    hover: '#EAEBFA',
  },
};

export const buttonStyle = css<BaseButtonProps>`
  outline: 0;
  cursor: pointer;
  align-items: center;
  display: inline-flex;
  text-decoration: none;
  box-sizing: border-box;
  justify-content: center;
  padding: ${spacing[3]} ${spacing[6]};
  width: ${({ fullWidth = false }) => fullWidth && '100%'};

  &:focus:not(:active) {
    border: 1px solid ${colors.white};
    box-shadow: 0 0 4px ${colors.actionPlain};
  }

  ${({ styling = 'primary', theme }) => {
    const { borderRadius, text } = theme?.button || {};
    const { size, height, weight } = text || {};
    const { primary, letterSpacingMap } = theme?.typography || {};
    const { text: textColor, border, bg: bgColor } = theme?.button?.[styling] || {};
    const backGround = bgColor ?? colorMap[styling].bg;
    const isActionGradient = backGround === colors.action;
    const bgFallback = css`
      background-color: ${colors.actionPlain};
    `;

    return css`
      background: ${backGround};
      border-radius: ${borderRadius ?? '8px'};
      border: ${border ?? '1px solid transparent'};
      font-family: ${primary ?? typography.primary};
      color: ${textColor ?? colorMap[styling].text};
      font-size: ${size ?? typography.sizes.text.body};
      font-weight: ${weight ?? typography.weights.semiBold};
      line-height: ${height ?? typography.sizes.lineHeight.body};
      letter-spacing: ${letterSpacingMap?.button ?? 0};
      ${isActionGradient && bgFallback};
    `;
  }}

  ${({ disabled, styling = 'primary', theme }) => {
    const { hover } = theme?.button?.[styling] || {};
    const { bg, text, border } = theme?.button?.[styling]?.disabled || {};
    const disabledStyles = css<BaseButtonProps>`
      cursor: not-allowed;
      ${({ loading }) => {
        if (!loading) {
          return css`
            color: ${text ?? colors.greyDark};
            background: ${bg ?? colors.greyLightest};
            border: ${border ?? '1px solid transparent'};
          `;
        }
      }}
    `;
    const enabledStyles = css`
      &:active {
        opacity: 0.8;
        box-shadow: unset;
        border: 1px solid transparent;
      }
      &:hover {
        background: ${hover ?? colorMap[styling].hover};
      }
    `;
    return disabled ? disabledStyles : enabledStyles;
  }}
`;

// This wrapper is to prevent html attribute warnings. See: https://styled-components.com/docs/faqs#why-am-i-getting-html-attribute-warnings
const ButtonWrapper = ({ loading, fullWidth, ...props }: ButtonProps & { theme: AppTheme }) => <button {...props} />;

const StyledWrapper = styled(ButtonWrapper)`
  ${buttonStyle}
`;

const Button = ({ children, loading, styling = 'primary', disabled, ...rest }: ButtonProps) => {
  const theme = useThemeContext();
  const isLoading = styling !== 'link' ? loading : undefined;

  return (
    <StyledWrapper styling={styling} loading={isLoading} disabled={isLoading || disabled} {...rest} theme={theme}>
      {isLoading && (
        <>
          <Spinner styling={styling === 'primary' ? 'negative' : 'secondary'} size="small" /> {'\u00A0 '}
        </>
      )}
      {children}
    </StyledWrapper>
  );
};

export default Button;
