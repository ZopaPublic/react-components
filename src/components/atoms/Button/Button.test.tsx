import axe from '../../../../axe-helper';
import React from 'react';
import { render } from '@testing-library/react';
import Button from './Button';

describe('<Button />', () => {
  it('renders with no a11y violations', async () => {
    const { container } = render(<Button>Primary</Button>);
    expect(container.firstChild).toMatchSnapshot();
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });

  it('renders as "secondary"', () => {
    const { container } = render(<Button styling="secondary">Secondary</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders as "link"', () => {
    const { container } = render(<Button styling="link">Secondary</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const { container } = render(<Button disabled>Secondary</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders wide', () => {
    const { container } = render(<Button fullWidth>Secondary</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders loading as "primary"', () => {
    const { container } = render(<Button loading>Loading</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders loading as "secondary"', () => {
    const { container } = render(
      <Button styling="secondary" loading>
        Waiting
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders loading as "secondary"', () => {
    const { container } = render(
      <Button styling="secondary" loading>
        Waiting
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders loading with disabled attribute', () => {
    const { container } = render(<Button loading>Loading</Button>);
    expect(container.firstChild).toHaveAttribute('disabled');
  });

  it.each`
    type
    ${'submit'}
    ${'reset'}
    ${'button'}
  `('renders with the specified type: $type', ({ type }) => {
    const { getByText } = render(<Button type={type}>Foo</Button>);
    const target = getByText('Foo');
    expect(target.getAttribute('type')).toEqual(type);
  });
});
