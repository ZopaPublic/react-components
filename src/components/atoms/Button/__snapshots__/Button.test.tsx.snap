// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders as "link" 1`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: transparent;
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #21247F;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

.c0:active {
  opacity: 0.8;
  box-shadow: unset;
  border: 1px solid transparent;
}

.c0:hover {
  background: #EAEBFA;
}

<button
  class="c0"
  styling="link"
  theme="[object Object]"
>
  Secondary
</button>
`;

exports[`<Button /> renders as "secondary" 1`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: #E9EAFA;
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #21247F;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

.c0:active {
  opacity: 0.8;
  box-shadow: unset;
  border: 1px solid transparent;
}

.c0:hover {
  background: #EEEFFB;
}

<button
  class="c0"
  styling="secondary"
  theme="[object Object]"
>
  Secondary
</button>
`;

exports[`<Button /> renders disabled 1`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: linear-gradient(90deg,#4F5AD8 0%,#3B46C4 100%);
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  background-color: #3B46C4;
  cursor: not-allowed;
  color: #4A545E;
  background: #F7F7F7;
  border: 1px solid transparent;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

<button
  class="c0"
  disabled=""
  styling="primary"
  theme="[object Object]"
>
  Secondary
</button>
`;

exports[`<Button /> renders loading as "primary" 1`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: linear-gradient(90deg,#4F5AD8 0%,#3B46C4 100%);
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  background-color: #3B46C4;
  cursor: not-allowed;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

<button
  class="c0"
  disabled=""
  styling="primary"
  theme="[object Object]"
>
  <img
    aria-label="loading spinner"
    height="20px"
    src="spinner-negative.gif"
    width="20px"
  />
   
    
  Loading
</button>
`;

exports[`<Button /> renders loading as "secondary" 1`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: #E9EAFA;
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #21247F;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  cursor: not-allowed;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

<button
  class="c0"
  disabled=""
  styling="secondary"
  theme="[object Object]"
>
  <img
    aria-label="loading spinner"
    height="20px"
    src="spinner-action.gif"
    width="20px"
  />
   
    
  Waiting
</button>
`;

exports[`<Button /> renders loading as "secondary" 2`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: #E9EAFA;
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #21247F;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  cursor: not-allowed;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

<button
  class="c0"
  disabled=""
  styling="secondary"
  theme="[object Object]"
>
  <img
    aria-label="loading spinner"
    height="20px"
    src="spinner-action.gif"
    width="20px"
  />
   
    
  Waiting
</button>
`;

exports[`<Button /> renders wide 1`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  width: 100%;
  background: linear-gradient(90deg,#4F5AD8 0%,#3B46C4 100%);
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  background-color: #3B46C4;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

.c0:active {
  opacity: 0.8;
  box-shadow: unset;
  border: 1px solid transparent;
}

.c0:hover {
  background: linear-gradient(90deg,#3B46C4 0%,#2732B0 100%);
}

<button
  class="c0"
  styling="primary"
  theme="[object Object]"
>
  Secondary
</button>
`;

exports[`<Button /> renders with no a11y violations 1`] = `
.c0 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: linear-gradient(90deg,#4F5AD8 0%,#3B46C4 100%);
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  background-color: #3B46C4;
}

.c0:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

.c0:active {
  opacity: 0.8;
  box-shadow: unset;
  border: 1px solid transparent;
}

.c0:hover {
  background: linear-gradient(90deg,#3B46C4 0%,#2732B0 100%);
}

<button
  class="c0"
  styling="primary"
  theme="[object Object]"
>
  Primary
</button>
`;
